hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ConsensusTimeHook.java

```diff
@@ -0,0 +1,102 @@
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Class responsible for running any actions that need to happen at the end of
+ * transaction handling. The reason it's called a consensus time hook is because
+ * the actions are (possibly) triggered by checking the previous transaction's
+ * consensus time against the consensus time of the current transaction.
+ */
+@Singleton
+public class ConsensusTimeHook {
+    private static final Logger logger = LogManager.getLogger(ConsensusTimeHook.class);
+
+    private final EndOfStakingPeriodUpdater stakingCalculator;
+    private Instant consensusTimeOfLastHandledTxn;
+
+    @Inject
+    public ConsensusTimeHook(@NonNull final EndOfStakingPeriodUpdater stakingPeriodCalculator) {
+        this.stakingCalculator = stakingPeriodCalculator;
+    }
+
+    /**
+     * Processing hook to run at the end of each transaction. There are certain actions that need
+     * to be taken once we have a new consensus timestamp. Any such actions should be done here
+     *
+     * @param consensusTime the consensus timestamp of the latest transaction being processed
+     * @param context the {@code HandleContext} context of the transaction being processed
+     */
+    public void process(@NonNull Instant consensusTime, @NonNull final HandleContext context) {
+        if (consensusTimeOfLastHandledTxn == null
+                || consensusTime.getEpochSecond() > consensusTimeOfLastHandledTxn.getEpochSecond()
+                        && isNextPeriod(consensusTimeOfLastHandledTxn, consensusTime, context)) {
+            // Handle the daily staking distributions and updates
+            try {
+                stakingCalculator.updateNodes(consensusTime, context);
+            } catch (final Exception e) {
+                logger.error("CATASTROPHIC failure updating end-of-day stakes", e);
+            }
+
+            // Advance the last consensus time to the given consensus time
+            consensusTimeOfLastHandledTxn = consensusTime;
+        }
+    }
+
+    @VisibleForTesting
+    void setLastConsensusTime(@Nullable final Instant lastConsensusTime) {
+        consensusTimeOfLastHandledTxn = lastConsensusTime;
+    }
+
+    @VisibleForTesting
+    static boolean isNextPeriod(
+            @NonNull final Instant lastConsensusTime,
+            @NonNull final Instant consensusTime,
+            @NonNull final HandleContext handleContext) {
+        final var stakingPeriod =
+                handleContext.configuration().getConfigData(StakingConfig.class).periodMins();
+        if (stakingPeriod == DEFAULT_STAKING_PERIOD_MINS) {
+            return !inSameUtcDay(lastConsensusTime, consensusTime);
+        } else {
+            return getPeriod(consensusTime, stakingPeriod * MINUTES_TO_MILLISECONDS)
```

Comment:
Side note:
The getPeriod method in `com.swirlds.common.stream.LinkedObjectStreamUtilities` is a bit odd (and not well documented).
It's basically using the Epoch as the reference point, then getting the number of periods since epoch (more precisely, the integer division of total-ms/ms-per-period).  Not entirely sure that correctly matches an arbitrary staking period (should it be since genesis, for instance?), but I suspect we have previous implementation that makes the same assumptions...

Reply:
Agreed, I had to plug in a bunch of values to get any idea for how this function works. As for this code, yes, the previous implementation made the same assumptions.